homeassistant/components/sensor/recorder.py

```diff
@@ -241,18 +246,29 @@ def compile_statistics(
         hass, start - datetime.timedelta.resolution, end, [i[0] for i in entities]
     )
 
-    for entity_id, state_class, key in entities:
-        wanted_statistics = DEVICE_CLASS_OR_UNIT_STATISTICS[state_class][key]
+    for entity_id, state_class, device_class in entities:
+        wanted_statistics = DEVICE_CLASS_STATISTICS[state_class][device_class]
 
         if entity_id not in history_list:
             continue
 
         entity_history = history_list[entity_id]
-        unit, fstates = _normalize_states(entity_history, key, entity_id)
+        unit, fstates = _normalize_states(entity_history, device_class, entity_id)
 
         if not fstates:
             continue
 
+        # Check metadata
+        if old_metadata := statistics.get_metadata(hass, entity_id):
+            if old_metadata["unit_of_measurement"] != unit:
+                if entity_id not in WARN_UNSTABLE_UNIT:
+                    _LOGGER.warning(
+                        "The unit of %s does not match the unit of statistics",
```

Comment:
Can we include the old and new one? 

Expected kWh, found Wh

Reply:
Fixed